make metadata and tagging consistent (all selected)

phweyland opened the pull request:
following discussion on #10032

When a metadata action applies on all selected images, include those which are not visible (grouped and collapsed).

Notes:
- this may not exhaustive because sometimes a specific list of images is provided by the calling function. In that case this change has no effect.
- actions on file (delete, ... ) are applied on all images including not not visible
- actions of development are applied only on visible images.

A critical review is welcome.

TurboGit commented:
I've read the discussion... But I'm really not sure!

Why should a tag be applied to a non visible grouped image? The tag could describe a details on the picture that do not happen to be visible on all images in the group. Same for GPX offset.

For pure metadata (author, note, ...) it probably makes sense.

But at the end, don't we prefer having a consistency across dt, that is **only do something on a picture** if visible? Otherwise people will have to read and remember the rules, what is set if not visible and what is not!

I would prefer keeping the behavior as it is today. Of course, and as usual maybe I'll change my mind if many good arguments are proposed for changing this. Let me tag this PR as controversial to catch attention !

phweyland commented:
The rule given by @AlicVB sounds OK for me (and should be close to what happens today). But "only on visible selected" would be a fair rule too.
the issue here is that metadata sometimes apply on visible and most of the time on all selected, if I'm not wrong.
Whatever the chosen rule,  it should be consistent.

phweyland commented:
With this change, any metadata action applies only on visible (selected) images.

Note: file actions (remove, delete, ....) still applies on all selected.

AlicVB commented:
I'm not sure... That would mean that rating (for example) is not applied on the jpeg sidecar... We will break workflow here, for sure !
imho the main problem is that grouping images can have very different meanings, depending of users. We can have RAW+jpeg, duplicates, bracketted images, panos, ... and each cases don't need the sames treatments... It's ok (and needed) to act on all images for RAW+jpeg, but not for duplicates...

I don't really see an issue on this :(
Maybe the better we can do is to add a preference "apply metadatas changes on hidden grouped images" ?

phweyland commented:
> the main problem is that grouping images can have very different meanings,

Exact.

> Maybe the better we can do is to add a preference "apply metadatas changes on hidden grouped images" ?

Can be the best solution

In any case there some issue in tagging, metadata and geotagging where the same rule is not applied everywhere.

AlicVB commented:
> In any case there some issue in tagging, metadata and geotagging where the same rule is not applied everywhere.

completely agree !

phweyland commented:
I've introduced a setting "apply metadata on visible images only" (not pushed) but I get other inconsistencies:
- in lighttable, the number of selected images includes the collapsed images.
  - It is strange to get only the visible images modified (add a tag or a color label), i.e. less images than selected
  - With drag & drop, all images are dragged and updated  anyway (tag, geotag, ...)
  - In metadata_view, what should be considered ? Only visible (if checked) or all selected ?

So I don't see this as in improvement, and I come back to @AlicVB assumption (as today):
- metadata and files actions => all selected (including collapsed ones).
- export and dev => only visible.

If the user wants not to apply  a metadata to all selected, he needs expand the collapsed images first.

If we agree on this, I'll just fix the current inconsistencies (mainly in tagging and metadata).

AlicVB commented:
it's not really my assumption, but more an "historical" things : that's how it was working for rating, colorlabels, then for tags, ... I've just "formalized" it a little bit ;)
imho it's fine, but as always, my workflow is not the only one !

that said, iiuc there's "just" some corner cases to fix so there's no more inconsistency, right ? in that case, I would say that the current situation is not so bad, considering I don't remember to have seen an issue about this lately...

phweyland commented:
@TurboGit, thoughts ?

TurboGit commented:
@phweyland : Ok, let's go with that, I have no string opinion even if I lean a bit forward uniformity (always do action on visible images).

phweyland commented:
>  I lean a bit forward uniformity (always do action on visible images).

I think there is way for that. To have a configuration point (only visible or all selected) and let dt_view_get_images_to_act_on() manage it (removing its parameter `only_visible`). Then all information, number of selected and applied operation would be always consistent. That could turn the code a bit stronger too.

AlicVB commented:
> To have a configuration point (only visible or all selected) and let dt_view_get_images_to_act_on() manage it (removing its parameter `only_visible`). Then all information, number of selected and applied operation would be always consistent. That could turn the code a bit stronger too.

As said earlier, I think the better way is to propose a preference to manage that. And yes, managing it at the `dt_view_get_images_to_act_on()` level can solve lot of issues, as everything should normally use this function.

Maybe we should then change the "only_visible" boolean argument for a 3 states one like "only_visible" "all images" "user_defined" because there will be cases where we don't want to let the choice to the user (dev actions or remove/delete for example)

phweyland commented:
> Maybe we should then change the "only_visible" boolean argument for a 3 states one like "only_visible" "all images" "user_defined" because there will be cases where we don't want to let the choice to the user (dev actions or remove/delete for example)

I don't understand the purpose of "user_defined". Could you elaborate a bit more please ?

AlicVB commented:
Yes, in fact I'm not sure we can completely remove the "only_visible" argument, because in some cases we will need to have exactly either the visible images either all images (dev actions, export, remove/delete, ...)
so to integrate the user preference here, we will need a 3 states argument : 
- only_visible = get only visible images, whatever the pref
- all_images = get all images whatever the pref
- user_defined = respect the pref

I hope I'm more clear

phweyland commented:
Thanks, I see now. You are talking about the parameter passed to `dt_view_get_images_to_act_on()`. That's another idea but I'm afraid that won't simplify the code, and will let visible some discrepancy between the displayed number of selected images and the number of images on which the action is applied.

While, letting `dt_view_get_images_to_act_on()` playing alone with the preference, everything is consistent, and, if that's your preference you can export the collapsed images or apply dev actions on them as well, or on the opposite, copy, move or delete only visible files. But as Pascal suggests, the action is always applied on the displayed number of selected images.

BTW: just to be sure, dev action is an action in lighttable like copy history, read xmp,... right ?

AlicVB commented:
> While, letting `dt_view_get_images_to_act_on()` playing alone with the preference, everything is consistent

no doubt about consistency, then, but I guess that most people will want to have inconsistent behavior ;)
think about the current state (which is not so problematic considering no one really object about it). With your proposal, that won't be possible anymore (or I have miss something)



> BTW: just to be sure, dev action is an action in lighttable like copy history, read xmp,... right ?

yes, everything related to image development including export, etc...

Anyway, I really don't think it's a good idea to let all that dev action be applied on hidden images. Think about "discard history" for example. I don't imagine a usecase where a user want to discard history of all grouped images, but with images grouped : 
- RAW+jpeg : no need as jpeg is not really used for development
- duplicates : as dev history and final rendering differ, it's quite dangerous to discard history blindly imho
- bracketing : mix of the 2 cases above : the group leader is likely to be the one used for developpment, and anyway the history will certainly differ between all the grouped images
- pano : maybe the only cases where we can imagine that you can have an uniformity of dev in all the images, but I'm even not sure : I'm far from a specialist, but I always need to tweak my images differently before the pano creation

But I may miss something here...

Maybe we can have a "middle state" ?
- only visible for all dev action
- preference for all metadatas actions

phweyland commented:
> I guess that most people will want to have inconsistent behavior ;)

Agreed. Agreed too there are a lot of different use cases.
Your examples make perfectly sense.
I guess that "only on visible" would also have counter examples (rating, ....).
The only trouble I see today is to see `x` images selected and to get `y` images treated (dev and export).

No problem. Currently most of the metadata actions are on `all selected`. I'll just try to fix the few exceptions.

phweyland commented:
#6025 related ?

phweyland commented:
make metadata and tagging consistent (all selected)
fix dt_tag_detach issue (group_on was not considered)
fix map view issue (grouped images not considered in cluster display)

Just the minimum.
